File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/Publish.java
package com.github.dedis.popstellar.model.network.method;

import com.github.dedis.popstellar.model.Immutable;
import com.github.dedis.popstellar.model.network.method.message.MessageGeneral;
import com.github.dedis.popstellar.model.objects.Channel;

import java.util.Objects;

/** Publish a message on a channel */
@Immutable
public final class Publish extends Query {

  private final MessageGeneral message;

  /**
   * Constructor for a Publish
   *
   * @param channel name of the channel
   * @param id request ID
   * @param message message to publish
   * @throws IllegalArgumentException if any parameter is null
   */
  public Publish(Channel channel, int id, MessageGeneral message) {
    super(channel, id);
    if (message == null) {
      throw new IllegalArgumentException("Trying to publish a null message");
    }
    this.message = message;
  }

  /** Returns the message to publish. */
  public MessageGeneral getMessage() {
    return message;
  }

  @Override
  public String getMethod() {
    return Method.PUBLISH.getMethod();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    if (!super.equals(o)) {
      return false;
    }
    Publish publish = (Publish) o;
    return Objects.equals(getMessage(), publish.getMessage());
  }

  @Override
  public int hashCode() {
    return Objects.hash(super.hashCode(), getMessage());
  }

  @Override
  public String toString() {
    return "Publish{"
        + "id="
        + getRequestId()
        + ", channel='"
        + getChannel()
        + '\''
        + ", message="
        + message
        + '}';
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/Publish.kt
package com.github.dedis.popstellar.model.network.method

import com.github.dedis.popstellar.model.Immutable
import com.github.dedis.popstellar.model.network.method.message.MessageGeneral
import com.github.dedis.popstellar.model.objects.Channel
import java.util.Objects

/** Publish a message on a channel */
@Immutable
class Publish(channel: Channel?, id: Int, message: MessageGeneral?) : Query(channel, id) {
  /** Returns the message to publish. */
  val message: MessageGeneral

  /**
   * Constructor for a Publish
   *
   * @param channel name of the channel
   * @param id request ID
   * @param message message to publish
   * @throws IllegalArgumentException if any parameter is null
   */
  init {
    requireNotNull(message) { "Trying to publish a null message" }

    this.message = message
  }

  override val method: String
    get() = Method.PUBLISH.method

  override fun equals(other: Any?): Boolean {
    if (this === other) {
      return true
    }
    if (other == null || javaClass != other.javaClass) {
      return false
    }
    if (!super.equals(other)) {
      return false
    }
    val publish = other as Publish
    return message == publish.message
  }

  override fun hashCode(): Int {
    return Objects.hash(super.hashCode(), message)
  }

  override fun toString(): String {
    return "Publish{id=$requestId, channel='$channel', message=$message}"
  }
}
